components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java

```diff
@@ -1322,13 +1543,51 @@ public void updateScope(ScopeDTO updatedScope) throws IdentityOAuthAdminExceptio
     public boolean isScopeExist(String scope) throws IdentityOAuthAdminException {
 
         int tenantId = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId();
+        String tenantDomain = IdentityTenantUtil.getTenantDomain(tenantId);
+        Boolean scopeExists;
         try {
-            return OAuthTokenPersistenceFactory.getInstance().getScopeClaimMappingDAO().isScopeExist(scope, tenantId);
-        } catch (IdentityOAuth2Exception e) {
+            if (OrganizationManagementUtil.isOrganization(tenantId) &&
+                    Utils.isClaimAndOIDCScopeInheritanceEnabled(tenantDomain)) {
+                String organizationId = OAuthComponentServiceHolder.getInstance().getOrganizationManager()
+                        .resolveOrganizationId(tenantDomain);
+                OrgResourceResolverService orgResourceManagementService =
+                        OAuthComponentServiceHolder.getInstance().getOrgResourceResolverService();
+                scopeExists = orgResourceManagementService.getResourcesFromOrgHierarchy(
+                        organizationId,
+                        LambdaExceptionUtils.rethrowFunction(orgId -> retrieveScopeExistenceInHierarchy(scope, orgId)),
+                        new FirstFoundAggregationStrategy<>()
+                );
+                return scopeExists != null && scopeExists;
+            } else {
+                return OAuthTokenPersistenceFactory.getInstance().
+                        getScopeClaimMappingDAO().isScopeExist(scope, tenantId);
+            }
+        } catch (IdentityOAuth2Exception | OrganizationManagementException | OrgResourceHierarchyTraverseException e) {
             throw handleError("Error while inserting the scopes.", e);
         }
     }
 
+    /**
+     * Checks whether a scope exists in the given organization in the hierarchy.
+     * @param scope The name of the scope to be retrieved.
+     * @param orgId The organization id of the tenant for which the scope needs to be retrieved.
+     * @return an optional containing true if it exists and an empty optional if it doesn't.
+     * @throws IdentityOAuth2Exception If an error occurs when getting the tenant id or checking the scope existence.
+     * @throws IdentityOAuthAdminException If an error occurs when getting the tenant domain.
+     */
+    private Optional<Boolean> retrieveScopeExistenceInHierarchy(String scope, String orgId)
```

Comment:
Can't we just return false here instead of Optional Boolean?

Reply:
This method is passed as an argument to the `getResourcesFromOrgHierarchy` [1] method which expects the return type to be `Optional<T>`. Therefore, returning a boolean instead of Optional<Boolean> would not compile due to a syntax error. 

[1] the https://github.com/wso2-extensions/identity-organization-management/blob/main/components/org.wso2.carbon.identity.organization.resource.hierarchy.traverse.service/src/main/java/org/wso2/carbon/identity/organization/resource/hierarchy/traverse/service/OrgResourceResolverServiceImpl.java#L42C70-L42C117